src/mappings/Cache.ts

```diff
@@ -0,0 +1,104 @@
+import {SubstrateEvent} from "@subql/types";
+import {blockNumber, eventId, timestamp} from "./common";
+import {Balance, AccountId} from "@polkadot/types/interfaces";
+import {RewardDestination} from "@polkadot/types/interfaces/staking";
+
+// Due to memory consumption optimization `rewardDestinationByAddress` contains only one key
+let rewardDestinationByAddress: {[blockId: string]: {[address: string]: RewardDestination}} = {}
+let controllersByStash: {[blockId: string]: {[address: string]: string}} = {}
+
+export async function cachedRewardDestination(accountAddress: string, event: SubstrateEvent): Promise<RewardDestination> {
+    const blockId = blockNumber(event)
+    let cachedBlock = rewardDestinationByAddress[blockId]
+    
+    if (cachedBlock !== undefined) {
+        return cachedBlock[accountAddress]
+    } else {
+        rewardDestinationByAddress = {}
+        
+        let method = event.event.method
+        let section = event.event.section
+
+        const allAccountsInBlock = event.block.events
+            .filter(blockEvent => { 
+                return blockEvent.event.method == method && blockEvent.event.section == section
+            })
+            .map(event => { 
+                let {event: {data: [accountId, ]}} = event
+                return accountId
+            });
+
+        const payees = await api.query.staking.payee.multi(allAccountsInBlock);
```

Comment:
What if `allAccountsInBlock` is empty?

Reply:
There is no use cases when it is possible but let's handle